### ISSUE TITLE
No Scope of API data

### ISSUE THREAD

**Issue by yashrajbothra** [openemr/openemr#3512]
### Describe the bug

So if we pass any column from `patient_data` table in POST data of core patient API that field will get inserted regardless of any scope of the User. I mean there should be a limit of the field which are allowed to the user. User should not be able to modify every field in the Database.

### To Reproduce

Steps to reproduce the behaviour:
1. Call POST `http://localhost:8081/openemr/apis/api/patient`
2.  Pass this Data
```json
{
    "title": "Mr",
    "fname": "Foo",
    "pid" : "55",
    "mname": "",
    "lname": "Bar",
    "street": "456 Tree Lane",
    "postal_code": "08642",
    "city": "FooTown",
    "state": "FL",
    "country_code": "US",
    "phone_cell" :  "hellla",
    "pharmacy_id" : "12",
    "phone_contact": "123-456-7890",
    "DOB": "1992-02-02",
    "sex": "Male",
    "pubid" : "1223",
    "race": "",
    "ethnicity": ""
}
```
Note It contains `pharmacy_id` which is not in Scope as per [API Guide](https://github.com/openemr/openemr/blob/dc298fa604e4a920c38fa252cbaeda6b46b72afa/API_README.md#post-apipatient)

### Expected behavior
It should give Error for Unexpected data :<

### Suggestion 
First, that's the controller's job! The services should exposed everything possible. 
The service should validate data integrity for sure but, what gets to the service will depend on the API audience and their role and permissions. With some folks we may want to allow only reads while others, full access.
This is especially true for our core services where we need services data exposure to be fully exposed for tasks such as form management.
What happens when two years down the road someone writes a controller that wants to read/update a field you deem inappropriate now. Or a new field gets added to a table, do you want me calling you up to come patch hard coded fields in code?

In the case of building queries for database transactions, that function should be kept pure and simple with its task. 

You're right to question this though so, you're thinking like an engineer. But, buildInsertColumns() is designed to dynamically build, not qualify. That duty should happen before getting there. We probably need a permissions function where we can whitelist columns based on API audiences logged in role/scope. Now that is worthy of an issue...

_Originally posted by @sjpadgett in https://github.com/openemr/openemr/pull/3335_

**Comment by sjpadgett** [comment]
I'm glad you're looking into this Yash and will be important for all our APIs. So keep that in mind.
Really BaseService hasn't anything to do with this as this needs to be global in scope to entire API project. Thus deserves its own classes and placement in the workflow.
We also need a base service and controller class to extend for FHIR. Doesn't need to be an interface class but an abstract class would be appropriate.

I'm just throwing things out there but, this issue will give a good opportunity to straighten these classes out.

**Comment by bradymiller** [comment]
Important to note that currently the API does have access controls in place which are user based. The same user within OpenEMR could modify all the patient_data column elements.

Also important to note that these services are going to evolve. For example, at some point the patient service will incorporate the very flexible layouts engine and stray away from the base service mechanism.

**Comment by yashrajbothra** [comment]
Should we start whitelisting the fields which can be used In **C**R**UD** opreation? As I was working on the practitioner resource and if we want **C**R**UD** than there should be some scope otherwise someone might mess up with username or password which we strictly don't want :)

**Comment by bradymiller** [comment]
definitely. right now the base service doesn't not allow modification of pid and uuid. would make sense to expand this type of thing as needed for the specific services. Also need to ensure `id` itself can't be modified.

**Comment by yashrajbothra** [comment]
> definitely. right now the base service does not allow modification of pid and uuid. would make sense to expand this type of thing as needed for the specific services. Also need to ensure `id` itself can't be modified.

Either we can Whitelist or Blacklist some. I think blacklist would be easy if we can do that?

**Comment by bradymiller** [comment]
Agree with blacklist strategy.

**Comment by stephenwaite** [comment]
bump
